### ISSUE TITLE
Issues with NuGet packages "Microsoft.AspNetCore.Authentication.Cookies" and "Microsoft.AspNetCore.Authentication.OpenIdConnect"

### ISSUE THREAD

**Issue by bhlee90** [aspnet/Security#1223]
There seem to be issues with NuGet packages "Microsoft.AspNetCore.Authentication.Cookies" (1.1.2) and "Microsoft.AspNetCore.Authentication.OpenIdConnect" (1.1.2). After I logged in with my work or school account and logged out, I was ultimately redirected to "https://localhost:44353/signout-callback-oidc", but I should have been redirected to "https://localhost:44353/" because my "config.json" is the following:

```
{
  "AzureAd": {
    "ClientId": "<OMITTED>",
    "AadInstance": "https://login.microsoftonline.com/{0}{1}", // This is the public instance of Azure AD
    "PostLogoutRedirectUri": "https://localhost:44353/"
  }
}
```

The following is my debug log:

```
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/Account/Login  
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) with arguments ((null)) - ModelState is Valid
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectMiddleware:Information: AuthenticationScheme: OpenIdConnect was challenged.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) in 3.8192ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 7.5053ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 POST http://localhost:44353/signin-oidc application/x-www-form-urlencoded 1634
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies signed in.
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 9.4364ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/  
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Cookies.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) with arguments ((null)) - ModelState is Valid
Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor:Information: Executing ViewResult, running view at path /Views/Home/Index.cshtml.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) in 7.4513ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 18.1401ms 200 text/html; charset=utf-8
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/Account/LogOff  
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Cookies.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.LogOff (WebApp-OpenIDConnect-DotNet) with arguments ((null)) - ModelState is Valid
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies was successfully authenticated.
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectMiddleware:Information: AuthenticationScheme: OpenIdConnect signed out.
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies signed out.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.AccountController.LogOff (WebApp-OpenIDConnect-DotNet) in 5.996ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 11.231ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/signout-callback-oidc  
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 2.1575ms 200 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/signout-callback-oidc  
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 6.9259ms 200 
```

For an explanation of where I got "44353" and "config.json" from, check out this sample [project](https://github.com/Azure-Samples/active-directory-dotnet-webapp-openidconnect-aspnetcore-v2). Before I updated "Microsoft.AspNetCore.Authentication.Cookies" and "Microsoft.AspNetCore.Authentication.OpenIdConnect", I noticed that after logging in and out of my work or school account, I was redirected to "https://localhost:44353", as expected. I was using 1.0.2. Here is my debug log:

```
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/Account/Login  
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) with arguments () - ModelState is Valid
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.5\System.IO.Compression.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectMiddleware:Information: AuthenticationScheme: OpenIdConnect was challenged.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) in 523.829ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 534.4578ms 302 
The thread 0x6194 has exited with code 0 (0x0).
The thread 0x94c has exited with code 0 (0x0).
The thread 0x9c4 has exited with code 0 (0x0).
The thread 0x6720 has exited with code 0 (0x0).
The thread 0x5d50 has exited with code 0 (0x0).
The thread 0x67a8 has exited with code 0 (0x0).
The thread 0x4fc4 has exited with code 0 (0x0).
The thread 0x16b0 has exited with code 0 (0x0).
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 POST http://localhost:44353/signin-oidc application/x-www-form-urlencoded 1698
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.5\System.Security.Cryptography.Encoding.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.5\System.Security.Cryptography.Csp.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.5\System.Security.Cryptography.Cng.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies signed in.
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 194.7283ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/  
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Cookies.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) with arguments () - ModelState is Valid
Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor:Information: Executing ViewResult, running view at path /Views/Home/Index.cshtml.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) in 10.6259ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 24.659ms 200 text/html; charset=utf-8
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/Account/LogOff  
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Cookies.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.LogOff (WebApp-OpenIDConnect-DotNet) with arguments () - ModelState is Valid
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies was successfully authenticated.
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectMiddleware:Information: AuthenticationScheme: OpenIdConnect signed out.
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies signed out.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.AccountController.LogOff (WebApp-OpenIDConnect-DotNet) in 27.7683ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 33.1758ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/signout-callback-oidc  
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 5.438ms 404 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/  
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) with arguments () - ModelState is Valid
Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor:Information: Executing ViewResult, running view at path /Views/Home/Index.cshtml.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) in 9.2481ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 17.2842ms 200 text/html; charset=utf-8
```

Also, in both versions 1.1.2 and 1.0.2 of "Microsoft.AspNetCore.Authentication.Cookies" and "Microsoft.AspNetCore.Authentication.OpenIdConnect", there appears to be an issue with logout redirection when logging out of a Microsoft account. I got stuck in the URL https://login.microsoftonline.com/common/oauth2/v2.0/logoutsession. The page shows the messages "You signed out of your account" and "It's a good idea to close all browser windows." I should have been redirected to "https://localhost:44353/".

Here is the log for 1.1.2:

```
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/Account/Login  
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) with arguments ((null)) - ModelState is Valid
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Users\bhlee_000\.nuget\packages\system.security.cryptography.encoding\4.3.0\runtimes\win\lib\netstandard1.3\System.Security.Cryptography.Encoding.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Users\bhlee_000\.nuget\packages\system.runtime.numerics\4.3.0\lib\netstandard1.3\System.Runtime.Numerics.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Users\bhlee_000\.nuget\packages\system.io.compression\4.3.0\runtimes\win\lib\netstandard1.3\System.IO.Compression.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectMiddleware:Information: AuthenticationScheme: OpenIdConnect was challenged.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) in 924.8205ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 937.2419ms 302 
The thread 0x3470 has exited with code 0 (0x0).
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 POST http://localhost:44353/signin-oidc application/x-www-form-urlencoded 1602
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Users\bhlee_000\.nuget\packages\system.security.cryptography.csp\4.3.0\runtimes\win\lib\netstandard1.3\System.Security.Cryptography.Csp.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Users\bhlee_000\.nuget\packages\system.security.cryptography.cng\4.3.0\runtimes\win\lib\netstandard1.6\System.Security.Cryptography.Cng.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies signed in.
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 192.0568ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/  
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Cookies.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) with arguments ((null)) - ModelState is Valid
Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor:Information: Executing ViewResult, running view at path /Views/Home/Index.cshtml.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.HomeController.Index (WebApp-OpenIDConnect-DotNet) in 5.9292ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 16.696ms 200 text/html; charset=utf-8
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/Account/LogOff  
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Cookies.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.LogOff (WebApp-OpenIDConnect-DotNet) with arguments ((null)) - ModelState is Valid
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies was successfully authenticated.
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectMiddleware:Information: AuthenticationScheme: OpenIdConnect signed out.
Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware:Information: AuthenticationScheme: Cookies signed out.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApp_OpenIDConnect_DotNet.Controllers.AccountController.LogOff (WebApp-OpenIDConnect-DotNet) in 30.8551ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 37.6937ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44353/signout-callback-oidc  
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 9.7075ms 200 
```

**Comment by bhlee90** [comment]
One key difference that I noticed between the 1.1.2 and 1.0.2 logs is that the 1.1.2 log has the following line:

`Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) with arguments ((null)) - ModelState is Valid`

The 1.0.2 log has the following line:

`Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method WebApp_OpenIDConnect_DotNet.Controllers.AccountController.Login (WebApp-OpenIDConnect-DotNet) with arguments () - ModelState is Valid`

**Comment by PinpointTownes** [comment]
This is a deliberate (breaking) design change, explained in this "announcement": https://github.com/aspnet/Security/issues/1087.

**Comment by bhlee90** [comment]
Right, I read that [announcement](https://github.com/aspnet/Security/issues/1087) before I created this issue. I forgot to mention that I read it in my original comment.

However, I still don't understand why my web app does not redirect me back to https://localhost:44353/ after logging in and out of my work or school account. Also, after I log in and out of a Microsoft account,  I am not redirected anywhere. I am stuck on https://login.microsoftonline.com/common/oauth2/v2.0/logoutsession.

**Comment by Eilon** [comment]
@bhlee90 can you confirm that you've configured the logout redirect URL in the Microsoft portal for your app? It also isn't clear to us from the information you posted as to which URL the browser ends up "stuck" on. Is it stuck on your web site's oidc signout callback URL, or is it stuck on the microsoftonline.com logoutsession URL?

**Comment by bhlee90** [comment]
@Eilon Yes, I configured the logout redirect URL in the Microsoft portal for my app, if you mean this [website](https://apps.dev.microsoft.com/). Here is a [screenshot](https://1drv.ms/i/s!ApxZYGZ3WCe8pA2hctGYKRPAq-sT) of the "Redirect URLs" and "Logout URL" that I set for the app (dev environment).

In my app, after I log out of my **work or school** account, the browser gets stuck on https://192.168.1.9:44373/signout-callback-oidc. Here is a [screenshot](https://1drv.ms/i/s!ApxZYGZ3WCe8pA8DnX_7pbrJiYF4) of the page.

In my app, after I log out of my **Microsoft** account, the browser gets stuck on https://login.microsoftonline.com/common/oauth2/v2.0/logoutsession. Here is a [screenshot](https://1drv.ms/i/s!ApxZYGZ3WCe8pA6l8foTQ8TsfJ_l) of the page.

Here is a snippet of the code in my Startup.cs responsible for using cookie auth and OpenID Connect auth:

`           
            // Configure the OWIN pipeline to use cookie auth.
            app.UseCookieAuthentication(new CookieAuthenticationOptions());

            var azureadclientid = "";
            var azureadaadinstance = "";
            var azureadpostlogoutredirecturi = "";

            if (env.IsDevelopment()) {
                azureadclientid = Configuration["AzureAD:Development:ClientId"];
                azureadaadinstance = Configuration["AzureAd:Development:AadInstance"];
                azureadpostlogoutredirecturi = Configuration["AzureAd:Development:PostLogoutRedirectUri"];
            } else {
                azureadclientid = Configuration["AzureAD:Production:ClientId"];
                azureadaadinstance = Configuration["AzureAd:Production:AadInstance"];
                azureadpostlogoutredirecturi = Configuration["AzureAd:Production:PostLogoutRedirectUri"];
            }


            // Configure the OWIN pipeline to use OpenID Connect auth.
            app.UseOpenIdConnectAuthentication(new OpenIdConnectOptions
            {
                ClientId = azureadclientid,
                Authority = string.Format(CultureInfo.InvariantCulture, azureadaadinstance, "common", "/v2.0"),
                ResponseType = OpenIdConnectResponseType.IdToken,
                PostLogoutRedirectUri = azureadpostlogoutredirecturi,
                Events = new OpenIdConnectEvents
                {
                    OnRemoteFailure = RemoteFailure,
                    OnTokenValidated = TokenValidated
                },
                TokenValidationParameters = new TokenValidationParameters
                {
                    // Instead of using the default validation (validating against
                    // a single issuer value, as we do in line of business apps), 
                    // we inject our own multitenant validation logic
                    ValidateIssuer = false,

                    NameClaimType = "name"
                }
            });
`

Again, here is the section in my appsettings.json file that has the configuration values of my dev env for Azure AD:

`
  "AzureAd": {
    "Development": {
      "ClientId": "<OMITTED>",
      "AadInstance": "https://login.microsoftonline.com/{0}{1}", // This is the public instance of Azure AD
      "PostLogoutRedirectUri": "https://192.168.1.9:44373/"
    },
`

**Comment by Tratcher** [comment]
Why are you using an IP address in PostLogoutRedirectUri? On a local machine you should use localhost. On a server you should use the host name.

**Comment by bhlee90** [comment]
@Tratcher The reason that I am using an IP address is that I wanted to test my web app on other devices besides the host. I am also using SSL encryption in the app.

I believe if I change the App URL to https://localhost:44373/, I won't be able to access the encrypted version of the website on another device, from my knowledge.

**Comment by Tratcher** [comment]
Are you using IIS Express? It's only setup to listen on localhost unless you run it as admin and override the host. And if you're going to do that why not use the machine name? Also note that https requires the host name to match the certificate, and IIS Express's cert is only for localhost.

**Comment by bhlee90** [comment]
Yes, I am using IIS Express. I don't know how to configure the web app to be accessible by a device besides localhost without configuring IIS Express to listen on an IP address. If I make it listen on localhost, I won't be able to access my website on another device.

I did not have to run IIS Express as an administrator, either. I have the following bindings in the config file in my project folder "C:\Users\bhlee_000\Documents\Visual Studio 2017\Projects\test.globalwarming.center\.vs\config" for IIS Express:

`
        <bindings>
			<binding protocol="https" bindingInformation="*:44358:localhost" />
            <binding protocol="https" bindingInformation="*:44358:192.168.1.9" />
        </bindings>
`

I also created and installed the SSL certificates for 192.168.1.9. I created it in PowerShell and imported it into the "Certificates" folder under "Trusted Root Certification Authorities".

Plus, I added the urlacl and the sslcert using netsh in PowerShell. I even had to copy the Thumbprint of the certificate that I installed and paste it as a parameter to add the sslcert.

Microsoft Edge recognizes my connection to the website https://192.168.1.9:44358 as encrypted. However, Google Chrome does not believe that my connection is secure. That is another matter that I don't want to bother resolving.

**Comment by Eilon** [comment]
Tagging @blowdart because he might have an idea on why this isn't working in Chrome. Something about SAN entry.

**Comment by blowdart** [comment]
Chrome changed how they validate self signed certificates; https://bugs.chromium.org/p/chromium/issues/detail?id=308330

So you need to ensure your localhost cert includes a SAN with the DnsName parameter in Powershell; like;

```PowerShell
$certificate = New-SelfSignedCertificate `
    -Subject localhost `
    -DnsName localhost `
    -KeyAlgorithm RSA `
    -KeyLength 2048 `
    -NotBefore (Get-Date) `
    -NotAfter (Get-Date).AddYears(5) `
    -CertStoreLocation "cert:CurrentUser\My" `
    -FriendlyName "IIS Express Development Certificate" `
    -HashAlgorithm SHA256 `
    -KeyUsage DigitalSignature, KeyEncipherment, DataEncipherment `
    -TextExtension @("2.5.29.37={text}1.3.6.1.5.5.7.3.1") 
```

**Comment by bhlee90** [comment]
@blowdart Chrome still complains that my connection to the test website is not secure, even after specifying a SAN for DnsName.

Anyway, my issues with logout redirection still remain.

**Comment by blowdart** [comment]
Hmm, does chrome give any hint why it thinks it's insecure? It could be the IP address. IP addresses in certs are frowned on these days, so it might be barfing on that,.

**Comment by bhlee90** [comment]
@blowdart Not quite. Even when I reconfigure my website to be accessible via "localhost" and put "localhost" as the SAN for the cert, I still run into the issue of Chrome thinking that my connection to the site is not secure. I do not want to even figure out the reason for that at the moment.

The SSL cert problem can wait. I am still trying to figure out why my web app does not redirect, properly, after logout. I am afraid to test the post logout redirect feature on the production version since it does not work on the development version.

**Comment by Eilon** [comment]
@bhlee90 sorry this appears to be a bit outside of our knowledge area because this seems to be related to issues with IIS Express. We recommend instead trying this on IIS (not Express) and deploy the app that way. If you're able to try that out and then let us know what is/isn't working we can go from there.

**Comment by bhlee90** [comment]
@Eilon
@blowdart
@Tratcher

Wow, I switched everything to localhost, and Microsoft Edge and Google Chrome recognized my connection to my site as secure.

I initially tried adding the site to IIS and gave up after I ran into an error that said that there was an error performing this operation and included the filename web.config, when I tried to access "Authentication" in IIS. Then, I tried launching the website on IIS Express, again. To my surprise, my connection was secure in Edge and Chrome. I am going to document what I did for the folks who are reading this:

I checked "Enable SSL" and put the same URL generated after checking the box in the "App URL" field:

[Link to screenshot](https://1drv.ms/i/s!ApxZYGZ3WCe8pCqFUclDRm7gGp0i)

Then, I had to run these commands in Command Prompt:

`
C:\WINDOWS\system32>netsh http add urlacl url=https://localhost:44373/ user=everyone

URL reservation successfully added

`
`
C:\WINDOWS\system32>netsh http add urlacl url=https://192.168.1.9:44373/ user=everyone

URL reservation successfully added


`

I also had to run these commands in PowerShell:

`
PS C:\Users\bhlee_000> $certificate = New-SelfSignedCertificate `
       -Subject localhost `
       -DnsName localhost `
       -KeyAlgorithm RSA `
       -KeyLength 2048 `
       -NotBefore (Get-Date) `
       -NotAfter (Get-Date).AddYears(5) `
       -CertStoreLocation "cert:LocalMachine\My" `
       -FriendlyName "IIS Express Development Certificate" `
       -HashAlgorithm SHA256 `
       -KeyUsage DigitalSignature, KeyEncipherment, DataEncipherment `
       -TextExtension @("2.5.29.37={text}1.3.6.1.5.5.7.3.1")
`

`
PS C:\Users\bhlee_000> $certificate = New-SelfSignedCertificate `
          -Subject 192.168.1.9 `
          -DnsName 192.168.1.9 `
          -KeyAlgorithm RSA `
          -KeyLength 2048 `
          -NotBefore (Get-Date) `
          -NotAfter (Get-Date).AddYears(5) `
          -CertStoreLocation "cert:LocalMachine\My" `
          -FriendlyName "IIS Express Development Certificate" `
          -HashAlgorithm SHA256 `
          -KeyUsage DigitalSignature, KeyEncipherment, DataEncipherment `
          -TextExtension @("2.5.29.37={text}1.3.6.1.5.5.7.3.1")
`

In addition, I copied the generated certificates from "Personal\Certificates" to "Trusted Root Certification Authorities\Certificates".

[Link to screenshot](https://1drv.ms/i/s!ApxZYGZ3WCe8pCtNe1aaDXL99dxG)

I think I had to clear my cache before visiting my site. It was strange that it did not work the first time. I want to thank you all for your help.
